tests/unit/nupic/algorithms/knn_classifier_test.py

```diff
@@ -29,6 +29,34 @@
 
 class KNNClassifierTest(unittest.TestCase):
 
+ 
+  
+  
+  
+  def testSparsifyVector(self):
+    classifier = KNNClassifier(distanceMethod="norm", distanceNorm=2.0)
+    inputPattern = np.array([0, 1, 3, 7, 11], dtype=np.int32)
+    
+    # Untouched
+    outputPattern = classifier._sparsifyVector(inputPattern, doWinners=True)
+    self.assertTrue(np.array_equal(np.array([0, 1, 3, 7, 11], dtype=np.int32), outputPattern))
+    
+    classifier = KNNClassifier(distanceMethod="norm", distanceNorm=2.0, relativeThreshold=True,
+      sparseThreshold=.2)
+    outputPattern = classifier._sparsifyVector(inputPattern, doWinners=True)
+    self.assertTrue(np.array_equal(np.array([0, 0, 3, 7, 11], dtype=np.int32), outputPattern))
+    
+    classifier = KNNClassifier(distanceMethod="norm", distanceNorm=2.0, relativeThreshold=True,
+      sparseThreshold=.2, numWinners=3)
+    outputPattern = classifier._sparsifyVector(inputPattern, doWinners=True)
+    self.assertTrue(np.array_equal(np.array([0, 0, 3, 7, 11], dtype=np.int32), outputPattern))
+    
+    classifier = KNNClassifier(distanceMethod="norm", distanceNorm=2.0, relativeThreshold=True,
+      sparseThreshold=.2, doBinarization=True)
+    outputPattern = classifier._sparsifyVector(inputPattern, doWinners=True)
+    self.assertTrue(np.array_equal(np.array([0., 0., 1., 1., 1.], dtype=np.int32), outputPattern))
```

Comment:
You **do** realize this is about the 2nd time I've written any Python right? ;-)

So my confusion is this. The method we're testing returns a list/array of integers _**most**_ of the time then in the last condition (doBinarization) it switches the type to a list/array of floats. see here:

```Python
def _sparsifyVector(self, inputPattern, doWinners=False):

    # Do sparsification, using a relative or absolute threshold
    if not self.relativeThreshold:
      inputPattern = inputPattern*(abs(inputPattern) > self.sparseThreshold)
    elif self.sparseThreshold > 0:
      inputPattern = inputPattern * \
        (abs(inputPattern) > (self.sparseThreshold * abs(inputPattern).max()))

    # Do winner-take-all
    if doWinners:
      if (self.numWinners>0) and (self.numWinners < (inputPattern > 0).sum()):
        sparseInput = numpy.zeros(inputPattern.shape)
        
        # Don't consider strongly negative numbers as winners.
        sorted = inputPattern.argsort()[0:self.numWinners]
        sparseInput[sorted] += inputPattern[sorted]
        inputPattern = sparseInput

    # Do binarization
    if self.doBinarization:
      # Don't binarize negative numbers to positive 1.
      inputPattern = (inputPattern > self.binarizationThreshold).astype(float)

    return inputPattern
```

The line under the comment (#doBinarization) results in a list/array (<-- I have no clue which) - of floats.

To me any concern of types seems ridiculous? So please educate me and tell me what in this code should lead me to concern for types? :-)